File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution(object):
    def checkInclusion(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        
        n=len(s1)
        if n==1:
            return s1 in s2
        freq_s1={}
        for c in s1:
            freq_s1[c]=freq_s1.get(c,0)+1
        j=0
        window={}
        for i in range(len(s2)):
            if s2[i] in freq_s1:
                window[s2[i]]=window.get(s2[i],0)+1
            else:
                window.clear()
                j=i
            if i-j+1==n:
                if window==freq_s1:
                    return True
                if s2[j] in window:
                    window[s2[j]] -= 1
                    if window[s2[j]]==0:
                        del window[s2[j]]
                j+=1
        return False
        

File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        unordered_map<char,int> mp1,mp2;
        for(char c:s1)mp1[c]++;
        int n=s1.size();
        int j=0;
        for(int i=0;i<s2.size();i++){
            if(mp1.count(s2[i])){
                mp2[s2[i]]++;
            }else{
                mp2.clear();
                j=i;
            }
            if(i-j+1==n){
                if(mp1==mp2) return true;
                if(mp2.count(s2[j])){
                    mp2[s2[j]]--;
                    if(mp2[s2[j]]==0){
                        mp2.erase(s2[j]);
                    }
                }
                j++;
            }


        }
        return false;
    }
};